### PR TITLE
ICU-20744 Add VS Code configuration files for c++

### PR DESCRIPTION
**Pull request by younies**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20744
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [ ] Tests included
- [x] Documentation is changed or added

### CODE REVIEW

File: docs/ide4c/vscode/tasks.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+            // You can pass all the building arguments here.
+            "args": [
+                "all",
+                "--debug", // Enable debug mode.
+                "-j8" // Use 8 cores.
+            ]
+        },
+        {
+            "label": "buildTest", // for building test cases before running it.
+            "command": "make", // Could be any other shell command
```
Comment (sffc) [inline docs/ide4c/vscode/tasks.json]: `make tests` does this.  Better than cd into the directory.

File: docs/ide4c/vscode/tasks.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            "args": [
+                "-j16",
+                "-C",
+                "${workspaceFolder}/test/intltest"
+            ],
+            "type": "shell"            
+        },
+        {
+            "label": "runTestCases", // For excuting the test cases.
```
Comment (sffc) [inline docs/ide4c/vscode/tasks.json]: Suggestion: rename this to "intltest", and then add two more targets, "cintltst" and "iotest" for the other test suites.

File: docs/ide4c/vscode/tasks.json [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+            "command": "${workspaceFolder}/test/intltest/intltest", // Could be any other shell command
+            "args": [
```
Comment (sffc) [inline docs/ide4c/vscode/tasks.json]: Please give an example for how you would specify a single test.  For example,

```js
"args": [
    // Specify a specific test case here; for example:
    // "format/NumberTest"
]
```

File: docs/ide4c/vscode/tasks.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+                "all",
+                "--debug", // Enable debug mode.
+                "-j8" // Use 8 cores.
+            ]
+        },
+        {
+            "label": "buildTest", // for building test cases before running it.
+            "command": "make", // Could be any other shell command
+            "args": [
+                "-j16",
```
Comment (sffc) [inline docs/ide4c/vscode/tasks.json]: `-j16` is a bit much, no?  I think Andy found that `-j6` gave the best performance and anything more than `-j6` had rapidly diminishing returns.

File: docs/ide4c/vscode/tasks.json [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+                "-C",
+                "${workspaceFolder}/test/intltest"
+            ],
+            "type": "shell"            
+        },
+        {
+            "label": "runTestCases", // For excuting the test cases.
+            "command": "${workspaceFolder}/test/intltest/intltest", // Could be any other shell command
```
Comment (sffc) [inline docs/ide4c/vscode/tasks.json]: Is there a way to make this command depend on a previous one like `buildTest`?

File: docs/ide4c/vscode/launch.json [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+                {
+                    "name": "LD_LIBRARY_PATH", // Use "DYLD_LIBRARY_PATH" for mac osx.
+                    "value": "${workspaceRoot}/lib:${workspaceRoot}/stubdata:${workspaceRoot}/tools/ctestfw"
+                },
+                {
+                    "name": "ICU_DATA",
+                    "value": "${workspaceRoot}/data/out"
+                }
+            ],
+            "args": [ // Here you add wich test cases you want to run, leaving it empty will run all the test cases.
```
Comment (jefgen) [inline docs/ide4c/vscode/launch.json]: very minor nit: typo (wich -> which)

**Comment by jefgen** [comment]
Thanks for doing this @younies! :)

**Comment by sffc** [comment]
Please also add the copyright headers link the Travis build is asking you to do.  For Markdown, you can use HTML-style comments.

File: docs/ide4c/vscode/tasks.json (hunk 1 above)
Comment (younies) [reply docs/ide4c/vscode/tasks.json]: You are right, however, tests should be added as args

File: docs/ide4c/vscode/tasks.json (hunk 4 above)
Comment (younies) [reply docs/ide4c/vscode/tasks.json]: Okay, I will make it -j6. However, we could not proof it is the best. For example, using Core i9 process may differ. Therefore, I will leave a comment states that is just the number of cores.

File: docs/ide4c/vscode/tasks.json (hunk 5 above)
Comment (younies) [reply docs/ide4c/vscode/tasks.json]: Yes, in the launch, we request buildTests first

File: docs/ide4c/vscode/tasks.json (hunk 3 above)
Comment (younies) [reply docs/ide4c/vscode/tasks.json]: This is in the launch task.

**Comment by younies** [comment]
Sorry, I performed the squash early, I was testing something.

File: docs/ide4c/vscode/launch.json [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+             "rbbi/RBBITest" // Just for example.
+            ],
+            "stopAtEntry": true, // If you want to add a break point in the begining of the program (i.e. test cases).
```
Comment (sffc) [inline docs/ide4c/vscode/launch.json]: Why default to true instead of false?  It is kind-of confusing to step from the main function down into your test case.  I usually leave a breakpoint in my test case and let the code run until it gets there.

File: docs/ide4c/vscode/launch.json [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+                    "name": "LD_LIBRARY_PATH", // Use "DYLD_LIBRARY_PATH" for mac osx.
+                    "value": "${workspaceRoot}/lib:${workspaceRoot}/stubdata:${workspaceRoot}/tools/ctestfw"
+                },
+                {
+                    "name": "ICU_DATA",
+                    "value": "${workspaceRoot}/data/out"
+                }
+            ],
+            "args": [ // Here you add wich test cases you want to run, leaving it empty will run all the test cases.
+             "rbbi/RBBITest" // Just for example.
```
Comment (sffc) [inline docs/ide4c/vscode/launch.json]: Inconsistent indentation

File: docs/ide4c/vscode/launch.json [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            "logging": {
+                "trace": true,
+                "traceResponse": true,
+                "engineLogging": true,
+                "programOutput": true
+            },
+            "preLaunchTask": "buildTest"
+        },
+        {
+            "name": "cintltest Launch",
```
Comment (sffc) [inline docs/ide4c/vscode/launch.json]: it's `cintltst`, no "e"

File: docs/ide4c/vscode/launch.json [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+            "MIMode": "gdb", // Use "lldb" for mac osx.
+            "logging": {
+                "trace": true,
+                "traceResponse": true,
+                "engineLogging": true,
+                "programOutput": true
+            },
+            "preLaunchTask": "buildTest"
+        }
+    ]
```
Comment (sffc) [inline docs/ide4c/vscode/launch.json]: Do you want to add `iotest` as well?  (example argument: `/datadriv`)

File: docs/ide4c/vscode/launch.json [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+                {
+                    "name": "LD_LIBRARY_PATH", // Use "DYLD_LIBRARY_PATH" for mac osx.
+                    "value": "${workspaceRoot}/lib:${workspaceRoot}/stubdata:${workspaceRoot}/tools/ctestfw"
+                },
+                {
+                    "name": "ICU_DATA",
+                    "value": "${workspaceRoot}/data/out"
+                }
+            ],
+            "args": [ // Here you add wich test cases you want to run, leaving it empty will run all the test cases.
```
Comment (sffc) [inline docs/ide4c/vscode/launch.json]: I find it useful to have an example default test arg, because I forget whether to start with a "/" or not (you need the leading "/" for cintltst but not intltest).  Here is one:

`"/tsutil/cstrcase"`

File: docs/ide4c/vscode/README.md [hunk 12]
```diff
@@ -0,0 +1,7 @@
+# Configuring VS Code for ICU4C
+
+  - create `.vscode` folder in icu4c/source
```
Comment (markusicu) [inline docs/ide4c/vscode/README.md]: I prefer to work from one or even two folder levels higher up. For example, with config+make I always use out-of-source builds because I don't want to contaminate my source tree, and I often configure two or more variants (debug vs. release vs. hardcode-UTF-8...) in parallel into separate output folders.

Does this kind of thing make sense for VS Code?
Does it support out-of-source builds?

File: docs/ide4c/vscode/README.md [hunk 13]
```diff
@@ -0,0 +1,7 @@
...
+  - Copy `tasks.json` and `launch.json` files into `.vscode` folder
+
+NOTE
+ before you build `icu4c` from vs code. run `./runConfigureICU` command first.
```
Comment (markusicu) [inline docs/ide4c/vscode/README.md]: Please specify whether this has to be an in-source build, or else how to tell VS Code about the relationship between the source and output folders.

File: docs/ide4c/vscode/launch.json [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    "configurations": [
+        {
+            "name": "intltest Launch", 
+            "type": "cppdbg",
+            "request": "launch",
+            "targetArchitecture": "x64",
+            "program": "${workspaceFolder}/test/intltest/intltest",
+            "environment": [
+                {
+                    "name": "LD_LIBRARY_PATH", // Use "DYLD_LIBRARY_PATH" for mac osx.
```
Comment (markusicu) [inline docs/ide4c/vscode/launch.json]: The red highlighting here of // comments indicates that they are not supported in .json, right? Does VS Code handle them, or does one have to strip them while copying into the .vscode folder?

Comment (sffc) [reply docs/ide4c/vscode/launch.json]: Vscode ignores // comments.

File: docs/ide4c/vscode/launch.json (hunk 7 above)
Comment (younies) [reply docs/ide4c/vscode/launch.json]: because the debugging can start even there is no break-points have been set. However, I agree with you. I will make the default  value false.

File: docs/ide4c/vscode/launch.json (hunk 8 above)
Comment (younies) [reply docs/ide4c/vscode/launch.json]: done!

File: docs/ide4c/vscode/launch.json (hunk 9 above)
Comment (younies) [reply docs/ide4c/vscode/launch.json]: Yes, this is the name in icu4c/source/test

File: docs/ide4c/vscode/launch.json (hunk 11 above)
Comment (younies) [reply docs/ide4c/vscode/launch.json]: Done!

File: docs/ide4c/vscode/launch.json (hunk 14 above)
Comment (younies) [reply docs/ide4c/vscode/launch.json]: Yes, vscode ignores them. and I think those comments are pretty useful when in this place.

File: docs/ide4c/vscode/launch.json (hunk 11 above)
Comment (sffc) [reply docs/ide4c/vscode/launch.json]: Except that you need the leading `/` in `/tsutil/cstrcase`

File: docs/ide4c/vscode/launch.json (hunk 9 above)
Comment (sffc) [reply docs/ide4c/vscode/launch.json]: Please change "cintltest launch" to "cintltst launch"

File: docs/ide4c/vscode/launch.json (hunk 10 above)
Comment (sffc) [reply docs/ide4c/vscode/launch.json]: Please add `iotest` unless there's a reason you don't want to include it

File: docs/ide4c/vscode/README.md (hunk 12 above)
Comment (sffc) [reply docs/ide4c/vscode/README.md]: We should aim at supporting one "suggested" workflow with these pre-built IDE files; people can customize them if they would like to do so.

I personally prefer my workspace root to be icu4c/source, largely because it reduces 2 levels of indentation in the file selector tree in the left panel of vscode.

File: docs/ide4c/vscode/launch.json (hunk 6 above)
Comment (younies) [reply docs/ide4c/vscode/launch.json]: Done !

File: docs/ide4c/vscode/README.md (hunk 13 above)
Comment (younies) [reply docs/ide4c/vscode/README.md]: Done ... I added a link to the page that talks about build/test ICU. There is no place in master/docs that explain the building process yet.

File: docs/ide4c/vscode/launch.json (hunk 9 above)
Comment (sffc) [reply docs/ide4c/vscode/launch.json]: Ping

File: docs/ide4c/vscode/launch.json (hunk 11 above)
Comment (sffc) [reply docs/ide4c/vscode/launch.json]: Ping

File: docs/ide4c/vscode/launch.json (hunk 10 above)
Comment (sffc) [reply docs/ide4c/vscode/launch.json]: Ping

File: docs/ide4c/vscode/launch.json [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+                    "value": "${workspaceRoot}/lib:${workspaceRoot}/stubdata:${workspaceRoot}/tools/ctestfw"
+                },
+                {
+                    "name": "ICU_DATA",
+                    "value": "${workspaceRoot}/data/out"
+                }
+            ],
+            "args": [ // Here you add wich test cases you want to run, leaving it empty will run all the test cases.
+            	"tsutil/cstrcase" // Just for example.
+	    ],
```
Comment (sffc) [inline docs/ide4c/vscode/launch.json]: Please fix indentation in this file (you have mixed tabs and spaces)

File: docs/ide4c/vscode/launch.json (hunk 9 above)
Comment (younies) [reply docs/ide4c/vscode/launch.json]: done

File: docs/ide4c/vscode/launch.json (hunk 10 above)
Comment (younies) [reply docs/ide4c/vscode/launch.json]: done!

**Comment by sffc** [comment]
Approved.  @markusicu, if you have any more comments, can you share them soon so that we can get this 3-month-old PR checked in?  Thanks!

**Comment by jefgen** [comment]
This is unrelated to your change... but the red syntax highlighting on the JSON file was bothering me in the diff view...
So after some searching, it turns out that GitHub actually does have a way to support JSON with comments in terms of the syntax highlighting.

It looks like you can add the following to the `.gitattributes` file:
`*.json linguist-language=jsonc`

This will cause the syntax highlighter to use the "jsonc" format, which supports comments.

I wonder if we'd want to maybe add this to the PR -- or perhaps I should file a separate ticket for it?

**Comment by sffc** [comment]
> I wonder if we'd want to maybe add this to the PR -- or perhaps I should file a separate ticket for it?

New ticket sounds good.

**Comment by jefgen** [comment]
Thanks!
I filed a new ticket, and created a new PR here:  https://github.com/unicode-org/icu/pull/969
